[core] CPD: Add total number of tokens to XML reports

maikelsteneker opened the pull request:
## Describe the PR

This PR extends the XML output of CPD. This output produces `<duplication>` elements for each duplication that is encountered. In addition to this, it now also produces `<file>` elements for each file with the number of tokens in the file added to this format. For example:
```
<?xml version="1.0" encoding="UTF-8"?>
<pmd-cpd>
   <file filename="/home/maikel/CR28584/file1.cs" totalNumberOfTokens="102"/>
   <file filename="/home/maikel/CR28584/file2.cs" totalNumberOfTokens="102"/>
   <duplication lines="15" tokens="102">
...
```

Other output formats currently don't contain this information, but could easily be extended to add it if needed.

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: Compared to master:<br />
This...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/324fc0f076de40d89a25975252067261/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel reviewed (commented):
Thanks for the PR!

I'm not sure, whether it's a good idea to add this additional method to the existing interface CPDRenderer.... wdyt @oowekyala ?

Documentation-wise we should also update the example on https://pmd.github.io/latest/pmd_userdocs_cpd_report_formats.html#xml
Maybe it's even time to add a formal documentation of the format or even a schema?
Add these additional elements to the XML format shouldn't be a problem. E.g. the maven-pmd-plugin would just ignore this additional elements (I hope...).

Could you also describe the use case a bit? What is this information used for? The number of tokens is an indication of how big the file is. Or do you maybe want to calculate later on that e.g. file x contains 20% duplicated code?

The XML format is now rather flat. Should we maybe group the duplication inside a `<file>` element? That of course would render the whole format incompatible, but we could introduce complete new XML renderer with a different format...

maikelsteneker commented on pmd-core/src/test/java/net/sourceforge/pmd/cpd/CPDCommandLineInterfaceTest.java:
```diff
@@ -22,8 +26,16 @@
+
 public class CPDCommandLineInterfaceTest {
     private static final String SRC_DIR = "src/test/resources/net/sourceforge/pmd/cpd/files/";
+    private static final Map<String, Integer> NUMBER_OF_TOKENS = ImmutableMap.of(
```
That is a fair point. In this case, I had to adjust the test cases because they would break otherwise. Should we maybe move them to `XMLRendererTest`?

adangel replied:
Let's keep them here - the test the CLI interface of CPD, but also use the XML format.
I see, that you added a test in XMLRendererTest now, so it's fine with me.

oowekyala commented:
Thanks for the contribution @maikelsteneker 

> I'm not sure, whether it's a good idea to add this additional method to the existing interface CPDRenderer.... wdyt @oowekyala ?

This is indeed my main gripe with the code change. I think adding a method parameter is not the right way, if we want to extend the info available to renderers in the future, it would be more flexible to pack all parameters in an object (eg a "CpdReport"). I would rather introduce an interface CpdReportRenderer whose method takes a CpdReport and a Writer, and deprecate the old interface. This way we can add members to CpdReport later without breaking the renderer interface

I wouldn't add an AbstractCpdRenderer either. The renderer interfaces should remain simple and an abstract class hints otherwise. We can use an adapter pattern to convert from the old interface to the new one.

maikelsteneker commented:
@oowekyala Thank you for the feedback!

I think you raised some very good points, so I have reworked my solution to resemble your proposal more closely. That is, I have introduced a new interface, the XML renderer is the only one that is changed as a result and the remaining renderers use an adapter to remain functioning. I guess that in the PMD 7.x branch, all renderers should implement just the latest interface, with the older ones being removed.

@adangel @oowekyala Could you take another look to see if this new implementation is up to your standards? Thanks in advance for your efforts!

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDTask.java:
```diff
@@ -130,7 +131,8 @@ private void report(CPD cpd) throws ReportException {
         }
-        CPDRenderer renderer = createRenderer();
+        CPDReportRenderer renderer = createRenderer();
+        CPDReport report = new CPDReport(cpd.getMatches(), cpd.getNumberOfTokensPerFile());
```
I think it would be nicer to make a method on `CPD` that encapsulates that:
```suggestion
        CPDReport report = cpd.toReport();
```

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDReport.java:
```diff
@@ -0,0 +1,26 @@
+    private final Iterator<Match> matches;
+    private final Map<String, Integer> numberOfTokensPerFile;
+
+    public CPDReport(final Iterator<Match> matches, final Map<String, Integer> numberOfTokensPerFile) {
```
```suggestion
    CPDReport(final Iterator<Match> matches, final Map<String, Integer> numberOfTokensPerFile) {
```

maikelsteneker replied:
This change results in errors for me, since I can no longer access the `CPDReport` class from `CPDReportRenderer`. Do you think it's a problem if this class is publicly accessible?

oowekyala replied:
I just mean to hide the constructor, the class is still accessible, right?

maikelsteneker replied:
Ah, my bad, you're right!

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDConfiguration.java:
```diff
@@ -382,11 +413,22 @@ public void setRenderer(Renderer renderer) {
         this.renderer = null;
     }
 
+    public void setCPDReportRenderer(CPDReportRenderer renderer) {
```
```suggestion
    public void setRenderer(CPDReportRenderer renderer) {
```

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDConfiguration.java:
```diff
@@ -326,10 +349,18 @@ public Renderer getRenderer() {
         return cpdRenderer;
     }
 
+    public CPDReportRenderer getCPDReportRenderer() {
```
```suggestion
    CPDReportRenderer getRenderer() {
```

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPD.java:
```diff
@@ -207,11 +217,13 @@ public static StatusCode runCpd(String... args) {
                 System.out.println(arguments.getRenderer().render(cpd.getMatches()));
             } else {
-                arguments.getCPDRenderer().render(cpd.getMatches(), new BufferedWriter(new OutputStreamWriter(System.out)));
+                final CPDReport report = new CPDReport(cpd.getMatches(), cpd.numberOfTokensPerFile);
```
```suggestion
                final CPDReport report = cpd.toReport();
```

oowekyala commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPD.java:
```diff
@@ -59,6 +63,10 @@ public Iterator<Match> getMatches() {
         return matchAlgorithm.matches();
     }
 
+    public Map<String, Integer> getNumberOfTokensPerFile() {
```
This is now accessible from the report

```suggestion
    private Map<String, Integer> getNumberOfTokensPerFile() {
```

maikelsteneker replied:
I think the method can even be removed then; we can directly use the `numberOfTokensPerFile` field.

oowekyala reviewed (commented):
Thanks for doing these changes, just a couple of minor comments

adangel commented on pmd-core/src/test/java/net/sourceforge/pmd/cpd/XMLRendererTest.java:
```diff
@@ -184,6 +188,37 @@ public void testWithOneDuplicationWithColumns() throws IOException {
+            final NamedNodeMap attributes = file.getAttributes();
+            assertEquals("/var/Foo.java", attributes.getNamedItem("path").getNodeValue());
+            assertEquals("888", attributes.getNamedItem("totalNumberOfTokens").getNodeValue());
+        } catch (Exception e) {
```
You don't need this try-catch. If an exception occurs in the test, the test will fail automatically, including a nice stacktrace. Please remove.

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/XMLRenderer.java:
```diff
@@ -101,10 +107,31 @@ public String render(Iterator<Match> matches) {
         doc.appendChild(root);
 
+        final List<Map.Entry<String, Integer>> entries = new ArrayList<>(numberOfTokensPerFile.entrySet());
+        Collections.sort(entries, new Comparator<Map.Entry<String, Integer>>() {
+            @Override
+            public int compare(final Map.Entry<String, Integer> entry1, final Map.Entry<String, Integer> entry2) {
+                return entry1.getKey().compareTo(entry2.getKey());
+            }
+        });
```
We could move this sorting into CPDReport - when we specify, that the CPDReports returns a SortedMap and the report returns the files pre-sorted. Not sure, if this is actually useful or whether using SortedMap would be too restrictive for CPDReport.

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDReport.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public Map<String, Integer> getNumberOfTokensPerFile() {
+        return Collections.unmodifiableMap(numberOfTokensPerFile);
```
The unmodifiable map can be created once in the constructor, then we have a simple getter here.

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDReport.java:
```diff
@@ -0,0 +1,27 @@
+        this.numberOfTokensPerFile = numberOfTokensPerFile;
+    }
+
+    public Iterator<Match> getMatches() {
```
Since we are defining this new useful class: I find it rather unusual, to return an Iterator (I know, that's what CPD used to do)

Should we rather return a List here? In order to not change more files (CPD, MatchAlgorithm, ...), we could simply iterate through the matches in the constructor for now and create a new (unmodifiable) list in this class...

maikelsteneker replied:
I completely agree that a list would be preferable. My only reason against switching to a list would be that I would expect that to increase the memory requirement for CPD. I'm not sure if this is actually an issue though.

adangel replied:
I think memory-wise there will be no difference: The iterator needs to hold on internally to the original list anyway. The only difference is, that you can iterate only once - you have a one-shot report.

I'll try to make the actual `List<Match>` available here...

adangel reviewed (commented):
Thanks for the changes!

From my point of view the only thing we should decide about is: Should we return a `List<Matches>` instead of `Iterator<Matches>` in the new CPDReport?

My other comments are some minor things that can also be changed later.

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/renderer/CPDRendererAdapter.java:
```diff
@@ -0,0 +1,23 @@
+
+import net.sourceforge.pmd.cpd.CPDReport;
+
+public class CPDRendererAdapter implements CPDReportRenderer {
```
I'll mark this as internal API, since this is just a utility for us to avoid changing all renderers now (which we eventually have to do anyway).

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDConfiguration.java:
```diff
@@ -170,7 +175,12 @@ public void postContruct() {
-                setCPDRenderer(getCPDRendererFromString(getRendererName(), getEncoding()));
+                try {
+                    setRenderer(getCPDReportRendererFromString(getRendererName(), getEncoding()));
+                } catch (ClassCastException e) {
+                    // The renderer class configured is not using the new CPDReportRenderer interface...
+                    setCPDRenderer(getCPDRendererFromString(getRendererName(), getEncoding()));
+                }
```
This is not necessary, since getCPDReportRendererFromString will call getCPDRendererFromString already and use the adapter to convert old CPDRenderers into new CPDReportRenderers.

Consequently this means, that when you implement a new CPDReportRenderer you have to implement the old interface as well, otherwise there is no way to use your renderer...

I think, I'll reimplement this logic here to use instanceof checks rather than classcastexceptions and try each interface in turn.

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDConfiguration.java:
```diff
@@ -231,6 +245,15 @@ public static CPDRenderer getCPDRendererFromString(String name, String encoding)
         }
     }
 
+    public static CPDReportRenderer getCPDReportRendererFromString(String name, String encoding) {
```
I'll make this new method package private and mark the others as internal API - these are just utility methods to convert a name of a renderer to the actual renderer

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDConfiguration.java:
```diff
@@ -326,10 +349,18 @@ public Renderer getRenderer() {
         return cpdRenderer;
     }
 
+    CPDReportRenderer getCPDReportRenderer() {
```
This is package-private only, so I'll change the deprecation notices above to internal API. We can make it public API, once there is a use case

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cpd/CPDConfiguration.java:
```diff
@@ -382,11 +413,22 @@ public void setRenderer(Renderer renderer) {
         this.renderer = null;
     }
 
+    public void setRenderer(CPDReportRenderer renderer) {
```
I think, we should make this package-private for now as well. The only way to set a renderer is by name.

adangel reviewed (approved):
Thanks for the updated PR.

I'll change the things I've mentioned and merge this afterwards.
